### PR TITLE
[Spot] Let cancel interrupt the spot job (#1414)

### LINKED ISSUE #1146: [Spot] `sky spot cancel` does not interrupt PENDING or STARTING job
The current behavior of our code is that we will set the spot job status for PENDING jobs to STARTING first before setting it to CANCLLED, which can be a bit confusing.

### PR DESCRIPTION
**Pull request by Michaelvll**
Closes #1146

This PR is a resubmission of #1414, fixing the issue in #1431.

Tested:
- [x] `./tests/run_smoke_tests.sh test_spot_cancellation`
- [x] `./tests/run_smoke_tests.sh test_spot`
- [x] `./tests/run_smoke_tests.sh test_spot_gcp`
- [x] `./tests/run_smoke_tests.sh test_spot_recovery`
- [x] `./tests/run_smoke_tests.sh test_spot_recovery_multi_node`
- [x] `./tests/run_smoke_tests.sh test_spot_storage`
- [x] `tests/run_smoke_tests.sh test_spot_tpu`

### CODE REVIEW

File: sky/spot/controller.py [hunk 1]
```diff
@@ -167,40 +166,82 @@ def start(self):
...
+        except ray.exceptions.RayTaskError:
+            # When the controller task is cancelled, it will raise
+            # ray.exceptions.RayTaskError, which can be ignored,
+            # since the SpotUserCancelledError will be raised and
+            # handled later.
+            pass
 
 
 if __name__ == '__main__':
+    ray.init()
```
Comment (romilbhardwaj) [inline sky/spot/controller.py:244]: Can we move this `ray.init()` call to inside `start()`? That way any external module can also directly use the spot controller by simply calling `controller.start()`.

**Review by romilbhardwaj — COMMENTED** [review]
Thanks for the quick fix! Trying it out now.

File: sky/spot/controller.py (hunk 1 above)
Comment (Michaelvll) [reply sky/spot/controller.py]: Good point! Moved. Thanks!

File: tests/test_smoke.py [hunk 2]
```diff
@@ -854,6 +854,48 @@ def test_spot_recovery_multi_node():
...
+            'sleep 300',
+            f's=$(sky spot queue); printf "$s"; echo; echo; printf "$s" | grep {name} | head -n1 | grep "RUNNING"',
+            # Terminate the cluster manually.
+            (f'aws ec2 terminate-instances --region {region} --instance-ids $('
+             f'aws ec2 describe-instances --region {region} '
+             f'--filters Name=tag:ray-cluster-name,Values={name}-2* '
+             f'--query Reservations[].Instances[].InstanceId '
+             '--output text)'),
+            'sleep 50',
+            f's=$(sky spot queue); printf "$s"; echo; echo; printf "$s" | grep {name}-2 | head -n1 | grep "RECOVERING"',
```
Comment (romilbhardwaj) [inline tests/test_smoke.py:886]: Do we need to test recovery in this test? Isn't it already covered by `test_spot_recovery` and `test_spot_recovery_multi_node`? Maybe we can remove it in the interest of keeping our tests fast (while avoiding test duplication)

**Comment by romilbhardwaj** [comment]
I'm having some trouble having my spot task cancel in the middle of setup. My spot task does not cancel, and stays stuck at STARTING.

My YAML:
```
file_mounts:
  /covid:
    source: s3://fah-public-data-covid19-cryptic-pockets # or name: mybucket123
    mode: MOUNT

resources:
  cloud: aws

setup: |
  echo "running setup"
  sleep 10000
  echo "done"

run: |
  conda env list
  sleep 600
```

I then run `sky spot cancel 2` (This was job 2). That seems to succeed:
```
$ sky spot cancel 2
Cancelling managed spot jobs with IDs 2. Proceed? [Y/n]: Y
I 11-20 00:16:00 spot_utils.py:149] Cancelling jobs 2.
Job with ID 2 is scheduled to be cancelled.
```

However, it's been about 15 min since I did a spot cancel, but my job still seems to be running the setup phase. `sky spot queue` shows this:
```
$ sky spot queue

Fetching managed spot job statuses...
Managed spot jobs:
In progress jobs: 1 STARTING

ID  NAME             RESOURCES     SUBMITTED    TOT. DURATION  JOB DURATION  #RECOVERIES  STATUS
2   sky-f436-romilb  1x [CPU:0.5]  23 mins ago  23m 57s        -             0            STARTING
1   sky-ece9-romilb  1x [CPU:0.5]  26 mins ago  5m 45s         3m 9s         0            CANCELLED
(base) romilb@romilbx1yoga:/mnt/d/Romil/Berkeley/Research/sky-experiments/examples/perf/oom_test$ sky 
```

Controller logs seem to be stuck too:
```
(base) romilb@romilbx1yoga:/mnt/d/Romil/Berkeley/Research/sky-experiments/examples$ sky logs sky-spot-controller-a70f1c17 2
Tailing logs of job 2 on cluster 'sky-spot-controller-a70f1c17'...
I 11-20 00:17:31 log_lib.py:388] Start streaming logs for job 2.
INFO: Tip: use Ctrl-C to exit log streaming (task will not be killed).
INFO: Waiting for task resources on 1 node. This will block if the cluster is full.
INFO: All task resources reserved.
INFO: Reserved IPs: ['172.31.40.245']
(sky-f436-romilb pid=27580) 2022-11-19 23:58:15,822     INFO worker.py:1337 -- Connecting to existing Ray cluster at address: 172.31.40.245:6379...
(sky-f436-romilb pid=27580) 2022-11-19 23:58:15,827     INFO worker.py:1513 -- Connected to Ray cluster. View the dashboard at 127.0.0.1:8265 
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-19 23:58:16 storage.py:592] Storage type StoreType.S3 already exists.
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-19 23:58:16 spot_state.py:116] Launching the spot cluster...
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-19 23:58:16 optimizer.py:606] == Optimizer ==
(_run_controller pid=27703) I 11-19 23:58:16 optimizer.py:629]  23:58:16 optimizer.py:629] Estimated cost: $0.1 / hour
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-19 23:58:16 optimizer.py:685] Considered resources (1 node):
(_run_controller pid=27703) I 11-19 23:58:16 optimizer.py:714]  23:58:16 optimizer.py:714] ------------------------------------------------------------------------
(_run_controller pid=27703) I 11-19 23:58:16 cloud_vm_ray_backend.py:2848] Tip: to reuse an existing cluster, specify --cluster (-c). Run `sky status` to see existing clusters.
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-19 23:58:16 cloud_vm_ray_backend.py:988] To view detailed progress: tail -n100 -f /home/ubuntu/sky_logs/sky-2022-11-19-23-58-16-533946/provisio
n.log
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-19 23:58:17 cloud_vm_ray_backend.py:1248] Launching on AWS us-east-2 (us-east-2c,us-east-2b,us-east-2a)
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-19 23:59:34 log_utils.py:45] Head node is up.
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-20 00:00:37 cloud_vm_ray_backend.py:1083] Successfully provisioned or found existing VM.
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-20 00:00:38 cloud_vm_ray_backend.py:3038] Processing 1 storage mount.
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-20 00:00:38 backend_utils.py:1080] Mounting (to 1 node): s3://fah-public-data-covid19-cryptic-pockets -> /covid
(sky-f436-romilb pid=27580) (_run_controller pid=27703) I 11-20 00:00:40 cloud_vm_ray_backend.py:2118] Running setup on 1 node.
(sky-f436-romilb pid=27580) (_run_controller pid=27703) running setupanently added '3.21.34.139' (ECDSA) to the list of known hosts.
(sky-f436-romilb pid=27580) I 11-20 00:06:33 controller.py:229] Cancelling spot job 2...
``` 

Are you able to replicate this issue? Maybe I'm doing something wrong...

**Comment by Michaelvll** [comment]
Thank you very much for testing it out @romilbhardwaj! I can reproduce the problem. It is very weird that the `ray.cancel` fails to cancel the ray task. Let me take a deeper look into the problem.

File: tests/test_smoke.py (hunk 2 above)
Comment (Michaelvll) [reply tests/test_smoke.py:886]: This was intended to test that the spot job can be canceled immediately during the recovering. Do you think that makes sense?

Comment (romilbhardwaj) [reply tests/test_smoke.py:886]: Ahh that does make a lot of sense - good to keep this then!

**Comment by Michaelvll** [comment]
I replaced the `ray` based implementation with `multiprocess.Process` because:
1. After the refactorization, we don't need to handle the exceptions in the main process, which get around the following limitations we mentioned before:
  > multiprocessing.Process: We can start the process and check the liveness of the process with is_alive. We can cancel the process with os.kill(p.pid, signal.SIGINT). However, we cannot easily catch the exceptions happening inside the process from the main process. Additional codes are needed to do so, as mentioned here.
2. It is easier to control the signal sends to the controller process with `multiprocess.Process`, but the `ray.cancel` does not explicitly send the `SIGINT` to the controller task, causing the setup fail to be canceled problem above.

**Comment by Michaelvll** [comment]
Tested:
- [x] `./tests/run_smoke_tests.sh test_spot_cancellation`
- [x] `./tests/run_smoke_tests.sh test_spot`
- [x] `./tests/run_smoke_tests.sh test_spot_gcp`
- [x] `./tests/run_smoke_tests.sh test_spot_recovery`
- [x] `./tests/run_smoke_tests.sh test_spot_recovery_multi_node`
- [x] `./tests/run_smoke_tests.sh test_spot_storage`
- [x] `tests/run_smoke_tests.sh test_spot_tpu`

File: sky/spot/controller.py [hunk 3]
```diff
@@ -196,49 +207,47 @@ def _handle_signal(job_id):
...
-                ray.cancel(controller_task)
-                ray.get(controller_task)
-        except ray.exceptions.RayTaskError:
-            # When the controller task is cancelled, it will raise
-            # ray.exceptions.RayTaskError, which can be ignored,
-            # since the SpotUserCancelledError will be raised and
-            # handled later.
-            pass
+        if controller_process is not None:
+            logger.info('sending SIGTERM to controller process '
```
Comment (romilbhardwaj) [inline sky/spot/controller.py:243]: nit : capital S "Sending SIGTERM"

**Review by romilbhardwaj — APPROVED** [review]
Works well in my testing, lgtm!

**Comment by Michaelvll** [comment]
Tested:
- [ ] `tests/run_smoke_tests.sh` (except for #1428/#1408).

File: examples/managed_spot.yaml [hunk 4]
```diff
@@ -15,7 +15,7 @@ run: |
   import time
   import tqdm
 
-  for i in tqdm.trange(120):
+  for i in tqdm.trange(1200):
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply examples/managed_spot.yaml]: Ahh, no, that is a remnant. Fixed. Thanks!

**Pull request merged**